### DISCUSSION TITLE
Why does the generated Java models all have `additionalProperties`?

### DISCUSSION THREAD

**Discussion by pontus-eliason** [asyncapi/modelina#675] [Q&A] [answered]
Why does the generated Java models all have `additionalProperties`? How do I get rid of them. I can't seem to see a way.

**Comment by jonaslagoni** [comment, marked as answer]
I assume you are generating models either from AsyncAPI, OpenAPI, or raw JSON schema inputs.

The reason it is generated is that is what you allow in your input schema, by default in JSON Schema draft 7 `additionalProperties` are `true` allowing input JSON to contain properties beyond what you define within `{properties: {...}}`. 🙂 

Modelina generates the property so you have a way to set the additional properties at runtime.

To disable it, explicitly set the schema to `additionalProperties: false`

```json
{"type": "object", "properties": {...}, "additionalProperties": false, ... }
```

Whether we should support something like a global setting that disable this logic, I would love to hear your reasoning for it 🙂

**Comment by pontus-eliason** [comment]
Oh. Actually had no idea that was the default. Nothing I had encountered before from other generators. Thank you :)

**Reply by jonaslagoni** [reply]
You bet.

Yea, it is actually a fully intended feature, as I was missing this feature when I used other code generators. However, I can fully understand it can be confusing when it generates it since the reason is inherited by a specification 😅 

So thanks for raising the discussion 👍

**Reply by pontus-eliason** [reply]
Something seems odd though. The `additionalProperties` property generated for Java does not have a `@JsonAnyGetter` annotation, so the serialized object will not have the map entries of `additionalProperties` expanded. Is there something I am missing? This is even with `JAVA_JACKSON_PRESET` enabled.

**Reply by jonaslagoni** [reply]
Yea good question.

The reason why it does not have it is Jackson (as far as I know, if it does please let us know 😄) doesn't support unwrapping a property into JSON.

For example, given the following JSON Schema:

```json
{"additionalProperties": {"type": "boolean"}}
```
And given the model generated is called `TestModel`, if you use that model with Jackson:
```java
ObjectMapper objectMapper = new ObjectMapper();

TestModel model = new TestModel();
model.additionalProperties = new HashMap();
model.additionalProperties.put("test", true);

String modelAsString = objectMapper.writeValueAsString(model);
```

What you do not want to happen is the generated JSON is this:
```json
{"additionalProperties": {"test": true}}
```
But you rather want the "unwrapped" JSON:
```json
{"test": true}
```
Because this is what correctly validates against the JSON Schema it was generated from. Therefore some presets "ignore it" as it is simply not supported.

That is why we provide our own implementation for JSON unmarshalling and marshaling code: https://github.com/asyncapi/modelina/blob/master/docs/languages/Java.md#include-json-marshaling-and-unmarshaling-methods
